normalizes decorative and functional `img` `alt` attributes

cfm opened the pull request:
## Status

Ready for review

## Description of Changes

1. Resolves #5983
2. Resolves #5984 
3. Lays groundwork for #5985

It turned out to be easier to review all `alt`s application-wide, commit the changes by category or resolution, and then submit them together in one omnibus PR.  #5985 has some extra bits outstanding.

## Testing

1. Navigate through the Source and Journalist Interfaces with a screen-reader running and listen for the `alt` attributes in context.
2. Note that all `img` elements have `alt` attributes:
```sh-session
$ grep "<img" **/*.html | grep -cv "alt="
0
```
3. Note that all non-empty, non-customizable `alt` attributes are translated:
```sh-session
$ grep "<img" **/*.html | grep -v 'alt=""' | grep -v "organization_name" | grep -cv "gettext"
0
```

## Deployment

No special considerations.

## Checklist

### If you made changes to the server application code:

- [X] Linting (`make lint`) and tests (`make test`) pass in the development container

cfm committed:
adds empty ALTs for self-evidently decorative images

cfm committed:
adds empty ALTs for IMGs used decoratively in context

cfm committed:
adds explanatory ALTs to functional IMGs

cfm committed:
adds explanatory ALTs to status-indicator IMGs

cfm committed:
adds descriptive ALTs to content-bearing IMGs

cfm commented:
@KwadroNaut [asks](https://github.com/freedomofpress/securedrop/pull/5982#issuecomment-862793746) (thanks!):

> the session timeout broom icon: did you check how this is handled with Tails/Windows/... screen readers? I never recognize that icon as a broom, just know where to click ¯_(ツ)_/¯

I hear this read straight through, skipping the empty `alt=""` attributes.  I understand this to be the desired narration, since the "New Identity" label is included in the surrounding text and corresponds to the tooltip of that toolbar icon in Tor Browser itself.  Would you recommend a different approach here?

> Click the New Identity button in your Tor Browser's toolbar. This will clear your Tor Browser activity data on this device.

---

> the hover icon in the admin template: is that really the way to go with an empty alt tag when there is a tooltip otherwise?

I'd thought that the two `img`s (`{off,on}-hover`) would have their `alt`s read consecutively, in duplicate.  But this doesn't appear (sound!) to be the case based on how the screen-reader follows the CSS, so I've added `alt`s back in for both of them in the revised 88f2a87.

KwadroNaut commented:
> @KwadroNaut [asks](https://github.com/freedomofpress/securedrop/pull/5982#issuecomment-862793746) (thanks!):
> 
> > the session timeout broom icon: did you check how this is handled with Tails/Windows/... screen readers? I never recognize that icon as a broom, just know where to click ¯_(ツ)_/¯
> 
> I hear this read straight through, skipping the empty `alt=""` attributes. I understand this to be the desired narration, since the "New Identity" label is included in the surrounding text and corresponds to the tooltip of that toolbar icon in Tor Browser itself. Would you recommend a different approach here?
> 
How does your screenreader work when using Tor browser is my question: the corresponding tooltip in Tor Browser is the answer: it's good what you did!

zenmonkeykstop commented on securedrop/journalist_templates/admin.html:
```diff
@@ -28,20 +28,20 @@ <h1>{{ gettext('Admin Interface') }}</h1>
             <i title="{{ gettext('Delete user {username}').format(username=user.username) }}">
-              <img src="{{ url_for('static', filename='icons/trash.png') }}" class="off-hover" width="14" height="16" alt="{{ gettext('trash icon') }}">
-              <img src="{{ url_for('static', filename='icons/trash-hover-red.png') }}" class="on-hover" width="14" height="16" alt="{{ gettext('trash icon') }}">
+              <img src="{{ url_for('static', filename='icons/trash.png') }}" class="off-hover" width="14" height="16" alt="{{ gettext('Delete') }}">
```
For clarity and consistency with "Edit user" above, should these be "Delete User"?

zenmonkeykstop commented on securedrop/journalist_templates/admin.html:
```diff
@@ -28,20 +28,20 @@ <h1>{{ gettext('Admin Interface') }}</h1>
 	{% if user.id == g.user.id %}
         <td>
-          <img src="{{ url_for('static', filename='icons/trash-disabled.png') }}" width="14" height="16" alt="{{ gettext('trash icon disabled') }}">
+          <img src="{{ url_for('static', filename='icons/trash-disabled.png') }}" width="14" height="16" alt="{{ gettext('Delete disabled') }}">
```
As per below, I'd err on the side of making this more descriptive, like "User deletion disabled."

zenmonkeykstop commented on securedrop/journalist_templates/col.html:
```diff
@@ -30,23 +30,23 @@
                 <input type="checkbox" name="doc_names_selected" value="{{ doc.filename }}" class="doc-check">
                 <span title="{{ gettext('Read') }}" class="icon">
-                  <img src="{{ url_for('static', filename='icons/check.png') }}" class="icon" width="16" height="12" alt="{{ gettext('check icon') }}">
+                  <img src="{{ url_for('static', filename='icons/check.png') }}" class="icon" width="16" height="12" alt="{{ gettext('Check') }}">
```
The checkmark here indicates that it's a reply that has been "deleted" by the source, which means that it's removed from their view in the source interface tho it persists in the journalist interface. This is interpreted as it having been read by the source, so something like "reply read" or "reply viewed" might work better.

zenmonkeykstop commented on securedrop/journalist_templates/flashed.html:
```diff
@@ -4,11 +4,11 @@
       <div class="flash {{ category }}">
         {% if category == "notification" %}
-        <img src="{{ url_for('static', filename='i/flash-notification.png') }}" height="35" width="15">
+        <img src="{{ url_for('static', filename='i/flash-notification.png') }}" alt="" height="35" width="15">
```
Not all messages are prefixed with "Success!," "Warning!" etc, so these symbols are not purely decorative but indicate flash message type. As such it might be worth including descriptive alt text for them.

zenmonkeykstop commented on securedrop/journalist_templates/preferences_saved_flash.html:
```diff
@@ -6,11 +6,11 @@
           <div class="flash {{ category_status }}">
           {% if category_status == "success" %}
-              <img src="{{ url_for('static', filename='i/flash-success.png') }}" height="15" width="20">
+              <img src="{{ url_for('static', filename='i/flash-success.png') }}" alt="" height="15" width="20">
```
Same as per above for other flash message icons.

zenmonkeykstop commented on securedrop/source_templates/index.html:
```diff
@@ -74,7 +74,7 @@ <h2 class="welcome-text">
     <div class="bubble" id="security-setting-info" class="info">
       <ol>
-        <li>{{ gettext('Click the <img src="{icon}" alt="shield icon"> in the toolbar above').format(icon=url_for("static", filename="i/font-awesome/black/guard.svg"))  }}</li>
+        <li>{{ gettext('Click the <img src="{icon}" alt="&quot;Security Level&quot; button"> in the toolbar above').format(icon=url_for("static", filename="i/font-awesome/black/guard.svg"))  }}</li>
```
Does this need to be in quotes?

cfm replied:
Since I can't boldface this (very small amount of :-) ["[t]ext taken directly from a user interface"](https://docs.securedrop.org/en/stable/development/documentation_guidelines.html#glossary), I thought I'd quote it instead.  But I'm not set on this if you think that `Security Level button` is clear enough in context to be worth the simplicity.

zenmonkeykstop replied:
That's fair, it's a cromulent alternative

zenmonkeykstop reviewed (changes requested):
Some changes requested, but also I think there's some room for discussion on a couple of things that might negate said changes:
- Should icons in flash messages have descriptive alts (I lean towards yes because otherwise there's no guaranteed indicator of the message type/severity)?
- Should alt descriptions on action buttons just be a single verb like "Delete" or descriptive like "Delete Reply"? (Could go either way depending on whether context of action is clear)

eloquence commented:
@cfm This fully supersedes #5982, correct?

cfm commented:
On 21 Jun 2021, at 8:58, Erik Moeller wrote:

> @cfm This fully supersedes #5982, correct?

Yes, though not intentionally; #5982 just hadn’t received a 
maintainers’ review prior to my starting in here, both based on #5982 
and in consultation with @KwadroNaut.

cfm committed:
contextualizes actions described in functional ALTs

cfm commented:
Thanks for the review in https://github.com/freedomofpress/securedrop/pull/5996#pullrequestreview-687685824, @zenmonkeykstop.

> * Should icons in flash messages have descriptive alts (I lean towards yes because otherwise there's no guaranteed indicator of the message type/severity)?

I've mulled this over, and I think I lean towards a more-pedantic interpretation.  :-)  If there's a need for an indicator of the `category` outside of the `message` text, that suggests to me that messages are missing headings that would either:

1. be independent of the existing decorative (`alt=""`) images and then styled to be invisible on CSS `@media screen` (i.e., reserved for `@media speech`); or
2. replace the existing images, which would be refactored in CSS to be *purely* decorative on the `H{n}` element (i.e., not present as an `img` element at all).

Again, I want to concede that this *is* pedantic: both approaches will look the same in a browser and sound the same in a screen-reader.  But my instinct here is that the question of these `alt` attributes actually stands in for the structural questions of #5986 and #5987.

So, I propose to:

1. go ahead and add descriptive `alt`s here (changes forthcoming), so that this pull request will provide the desired narration for now; and then
2. flag these messages (and any similar cases) for consideration in #5986 and #5987.

How does this sound to you, @zenmonkeykstop?

> * Should alt descriptions on action buttons just be a single verb like "Delete" or descriptive like "Delete Reply"? (Could go either way depending on whether context of action is clear)

Good catch; thanks.  I'd planned to do another pass on these as part of #5985, to include information about *which* submission/reply/user will be subject to the action, but will go ahead now (changes forthcoming) for the cases you've flagged here.

cfm committed:
adds explanatory ALTs to flashed messages' category icons

These ALT attributes are effectively an accessibility shim, to
provide the desired narration of flashed messages pending a
more–semantically-minded refactoring in #5986 and #5987.

See discussion in
<https://github.com/freedomofpress/securedrop/pull/5996#issuecomment-866199638>.

cfm commented:
Re: https://github.com/freedomofpress/securedrop/pull/5996#issuecomment-863508238:

> I'd thought that the two `img`s (`{off,on}-hover`) would have their `alt`s read consecutively, in duplicate. But this doesn't appear (sound!) to be the case based on how the screen-reader follows the CSS, so I've added `alt`s back in for both of them in the revised [88f2a87](https://github.com/freedomofpress/securedrop/commit/88f2a87edafc1a34f7cc7567838998713a3898a1).

I've gotten different results on this question in different browsers, so I've re-marked this as draft while I test further.

cfm commented:
Re: https://github.com/freedomofpress/securedrop/pull/5996#issuecomment-866287505:
 
> > I'd thought that the two `img`s (`{off,on}-hover`) would have their `alt`s read consecutively, in duplicate. But this doesn't appear (sound!) to be the case based on how the screen-reader follows the CSS, so I've added `alt`s back in for both of them in the revised [88f2a87](https://github.com/freedomofpress/securedrop/commit/88f2a87edafc1a34f7cc7567838998713a3898a1).
> 
> I've gotten different results on this question in different browsers, so I've re-marked this as draft while I test further.

The duplication (in some cases, triplication) I'm hearing turns out to come not from the multiple `img[alt]` attributes but from (e.g.):

https://github.com/freedomofpress/securedrop/blob/b486c9acfde6a1216e96b3c6ad8dd32abdc8cc11/securedrop/journalist_templates/admin.html#L41-L46

1. the duplication of the `img[alt]` attributes in (e.g.) the `i[title]` attribute (which is apparently lately [considered harmful](https://silktide.com/blog/i-thought-title-text-improved-accessibility-i-was-wrong/)); and
2. idiosyncrasies in how different screen-readers (and their configurations) infer and narrate [ARIA groups](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_group_role).

So the changes proposed here stand in scope, and this behavior is more evidence in favor of the semantic refactoring ticketed in #5986 and #5987.

zenmonkeykstop reviewed (approved):
Changes look good to me, approving.
